File: minimarket-frontend/src/app/core/services/whatsapp.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PedidoWhatsApp, DetalleVentaRequest } from '../models/models';

@Injectable({ providedIn: 'root' })
export class WhatsAppService {
  private http = inject(HttpClient);
  private readonly baseUrl = '/api/whatsapp';

  listarTodos(): Observable<PedidoWhatsApp[]> {
    return this.http.get<PedidoWhatsApp[]>(`${this.baseUrl}/pedidos`);
  }

  listarPendientes(): Observable<PedidoWhatsApp[]> {
    return this.http.get<PedidoWhatsApp[]>(`${this.baseUrl}/pedidos/pendientes`);
  }

  listarPorTipo(tipo: string): Observable<PedidoWhatsApp[]> {
    return this.http.get<PedidoWhatsApp[]>(`${this.baseUrl}/pedidos/tipo/${tipo}`);
  }

  /** Conteo de pedidos por estado — para badge en navbar */
  contarPorEstado(estado: string = 'PENDIENTE'): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.baseUrl}/pedidos/count?estado=${estado}`);
  }

  marcarEnProceso(id: number): Observable<PedidoWhatsApp> {
    return this.http.put<PedidoWhatsApp>(`${this.baseUrl}/pedidos/${id}/en-proceso`, {});
  }

  atender(id: number, respuesta: string): Observable<PedidoWhatsApp> {
    return this.http.put<PedidoWhatsApp>(
      `${this.baseUrl}/pedidos/${id}/atender?respuesta=${encodeURIComponent(respuesta)}`,
      {}
    );
  }

  /** Enviar mensaje libre del operador (no cierra el pedido) */
  enviarMensaje(id: number, texto: string): Observable<PedidoWhatsApp> {
    return this.http.post<PedidoWhatsApp>(
      `${this.baseUrl}/pedidos/${id}/enviar`,
      { texto }
    );
  }

  /** Convierte un pedido WhatsApp en una Venta PENDIENTE */
  convertirAVenta(id: number, payload: {
    clienteId: number;
    cajeroId: number;
    productos: DetalleVentaRequest[];
  }): Observable<{ mensaje: string; ventaId: number; total: number; estado: string }> {
    return this.http.post<any>(`${this.baseUrl}/pedidos/${id}/convertir-venta`, payload);
  }
}

File: MiniMarket/src/main/java/edu/pe/cibertec/whatsapp/WhatsAppService.java
package edu.pe.cibertec.whatsapp;

import java.util.List;

import edu.pe.cibertec.entity.PedidoWhatsApp;
import edu.pe.cibertec.entity.Venta;

public interface WhatsAppService {

    void procesarMensajeEntrante(String numeroRemitente, String nombreRemitente, String mensaje);


    void enviarMensaje(String numeroDestino, String mensaje);

    List<PedidoWhatsApp> listarPendientes();

    List<PedidoWhatsApp> listarPorTipo(String tipo);

    List<PedidoWhatsApp> listarTodos();

    PedidoWhatsApp atender(Long id, String respuesta);

    PedidoWhatsApp marcarEnProceso(Long id);

    PedidoWhatsApp vincularVenta(Long pedidoId, Long ventaId);


    PedidoWhatsApp enviarMensajeOperador(Long pedidoId, String texto);

    long contarPorEstado(String estado);

    Venta convertirAVenta(Long pedidoId, Long clienteId, Long cajeroId,
                          java.util.List<edu.pe.cibertec.dto.DetalleVentaRequest> productos);
}
